helix-core/src/main/java/org/apache/helix/cloud/event/CloudEventHandler.java

```diff
@@ -0,0 +1,125 @@
+ * under the License.
+ */
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class implements a wrapper around multiple {@link CloudEventListener}, and arrange them by
+ * 1. PreEventHandlerCallback -> only one allowed
+ * 2. Unordered CloudEventListener list -> multiple allowed
+ * 3. PostEventHandlerCallback -> only one allowed
+ * to enable an easy management of event listeners and callbacks.
+ */
+public class CloudEventHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(CloudEventHandler.class.getName());
+  private List<CloudEventListener> _unorderedEventListenerList = new ArrayList<>();
+  private CloudEventListener _preEventHandlerCallback = null;
+  private CloudEventListener _postEventHandlerCallback = null;
+
+  /**
+   * Register an event listener to the event handler. The listeners will be triggered in parallel in events.
+   * @param listener
+   */
+  public void registerCloudEventListener(CloudEventListener listener) {
+    _unorderedEventListenerList.add(listener);
+  }
+
+  /**
+   * Unregister an event listener to the event handler.
+   * @param listener
+   */
+  public void unregisterCloudEventListener(CloudEventListener listener) {
+    _unorderedEventListenerList.remove(listener);
+  }
+
+  /**
+   * Register a callback to be triggered before all the event listeners,
+   * only one instance is allowed
+   * @param preEventHandlerCallback
+   */
+  public void registerPreEventHandlerCallback(CloudEventListener preEventHandlerCallback) {
+    _preEventHandlerCallback = preEventHandlerCallback;
+  }
+
+  /**
+   * Unregister the registered preEventHandlerCallback
+   * @param preEventHandlerCallback
+   */
+  public void unregisterPreEventHandlerCallback(CloudEventListener preEventHandlerCallback) {
+    _preEventHandlerCallback = null;
+  }
+
+  /**
+   * Register a callback to be triggered after all the event listeners,
+   * only one instance is allowed
+   * @param postEventHandlerCallback
+   */
+  public void registerPostEventHandlerCallback(CloudEventListener postEventHandlerCallback) {
+    _postEventHandlerCallback = postEventHandlerCallback;
+  }
+
+  /**
+   * Unregister the registered postEventHandlerCallback
+   * @param postEventHandlerCallback
+   */
+  public void unregisterPostEventHandlerCallback(
+      CloudEventListener postEventHandlerCallback) {
+    _postEventHandlerCallback = null;
+  }
+
+  /**
+   * Trigger the callback / listeners in order of
+   * 1. PreEventHandlerCallback
+   * 2. Unordered CloudEventListener list
+   * 3. PostEventHandlerCallback
+   * @param eventInfo the object contains any information about the incoming event
+   */
+  public void onPause(Object eventInfo) {
+    if (_preEventHandlerCallback != null) {
+      _preEventHandlerCallback.onPause(eventInfo);
+    }
+    _unorderedEventListenerList.forEach(listener -> listener.onPause(eventInfo));
```

Comment:
I remember parallel execution is using object.stream()?

Reply:
It should be `parallelStream()` or `stream().parallel()`, however parallelism is not guaranteed.

Reply:
Both stream() and forEach() are sequential, so I used parallelStream() instead. But I think sequential is actually fine, what we want to do here is not to guarantee a parallel processing, we just want to emphasize that the list is not ordered. What do you think? @junkaixue

Reply:
If it is not ordered, let's use parallelStream() since it will not be an efficient design for non order running in sequence.